SEP-38: distinguish sell/buy delivery methods for fee calculation (#989)
stellar/stellar-protocol: ecosystem/sep-0038.md

Name | Type | Description
-----|------|------------
`asset` | string | The [Asset Identification Format](#asset-identification-format) value.
`sell_delivery_methods` | array | (optional) Only for non-Stellar assets. An array of objects describing the methods a client can use to sell/deliver funds to the anchor. The method of delivery may affect the expiration and/or price provided in a [`POST /quote`](#post-quote) response.
`buy_delivery_methods` | array | (optional) Only for non-Stellar assets. An array of objects describing the methods a client can use to buy/retrieve funds from the anchor. The method of delivery may affect the expiration and/or price provided in a [`POST /quote`](#post-quote) response.
`country_codes` | array | (optional) Only for fiat assets. A list of [ISO 3166-1 alpha-3](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3) codes of the countries where the Anchor operates for fiat transactions.

> `delivery_methods` Object Schema, used for both parameters `sell_delivery_methods` and `buy_delivery_methods`.

Reviewer: It appears there is no `delivery_methods` anymore.
```suggestion
Object Schema for `sell_delivery_methods` and `buy_delivery_methods`.
```

Author: Good suggestion!

    },
    {
      "asset": "iso4217:BRL",
      "country_codes": ["BRA"],

>       "client_delivery_methods": [

Reviewer: If these are always for deposit, could we use the term `deposit_delivery_methods`, and similar for anchors, `withdrawal_delivery_methods`? That builds on language we're already using for these two operations, where as `client` and `anchor` requires the reader to infer that during deposit they are the deliverer and therefore need to use this field.

cc @JakeUrban As I know you originally suggested client/anchor here.

Reviewer 2: `deposit` and `withdraw` work too, but SEP-31 doesn't use that language. Maybe its still better than `client/anchor` since in SEP-31 the client is referred to as an anchor. I'm good with either 👍

Reviewer: `send` / `receive`?

Author: What about `buy`/`sell`? They will be used along with parameters like buy_asset, buy_amount, sell_asset and sell_amount

These prices are indicative. The actual price will be calculated at conversion time once the Anchor receives the funds from a User.

The prices returned include any margin that the provider may keep as a service fee, and this margin may vary depending on the directional flow of funds, amount, delivery method, country code, or other factors.

#### Request

Name | Type | Description
-----|------|------------
`sell_asset` | string | The asset you want to sell, using the [Asset Identification Format](#asset-identification-format).
`sell_amount` | string | The amount of `sell_asset` the client would exchange for each of the `buy_assets`.

> `sep` | string | The name of the SEP that will be executing the quoted exchange. As of now, the ones that can make use of RFQ are `SEP-6` or `SEP-31`.

Reviewer: I'm hesitant for us to include a SEP name in here. Can we include an `intent` or similar instead? For example, `"intent": "transfer"` for the transfer API SEPs SEP-6/24, `"intent": "cross_border_payment"` for SEP-31. There are probably better names.

Reviewer: @JakeUrban wondering what your thoughts are on this.

Author: Regarding `"intent"` vs `"sep"`, I think both have downsides: in the future, we could end up having different protocols for "cross_border_payment" or different kinds of quote-related actions inside the same SEP, so both `&sep=6` or `&intent=cross_border_payment` could become insufficient to describe the intended action. TBH I like the "intent" key for it being more generic and easier to expand, but I also like the values "sep6" and "sep31" as they give a precise definition of what it'll be used for and they're also easier to expand with edge cases.

---

> `delivery_method` is only used for off-chain --> on-chain (i.e. deposit) flows, so we may need another attribute for the opposite flow, describing how funds are delivered to the user.

@JakeUrban: maybe we could tweak the description of delivery_methods a bit so it can be used for both buy_asset and sell_asset:
* Update the description of delivery_methods to say `An array of objects describing the methods a client can use to deliver funds to or withdraw funds from the anchor.`
* Update the `/quote`, `/price` and `/prices` request parameters to have `buy_delivery_method` and `sell_delivery_method` as parameters that should be used when buy_asset or sell_asset are non-Stellar assets.

Reviewer 2: > Update the description of delivery_methods to say An array of objects describing the methods a client can use to deliver funds to or withdraw funds from the anchor.

I think this approach would make the spec more confusing and make it easier to pass semantically invalid `delivery_method` values. For example, a client could potentially pass `cash_pickup` for an off-chain --> on-chain transaction, when the appropriate value would've been `cash_dropoff`.

From a design perspective, the methods a client uses to deliver funds to the anchor are disjoint from the methods an anchor supports for delivering funds to the client/user. Why describe both sets of methods under the same object?

What if we changed the existing `delivery_methods` object & parameter to be `client_delivery_method` and added a `anchor_delivery_method` object & parameter? So `/quote`, `/price`, and `/prices` requests would include a `client_delivery_method` if it was an off-chain --> on-chain exchange and an `anchor_delivery_method` if it was an on-chain --> off-chain exchange.

Author: @JakeUrban: I applied your suggestion in https://github.com/stellar/stellar-protocol/pull/989/commits/533d52fa85965b55d87ec71b3388c0e0292278aa. 

Q: should this trigger a major version increment from `1.2.0` to `2.0.0` or are we ok going to 1.3.0?

Reviewer 2: > Q: should this trigger a major version increment from 1.2.0 to 2.0.0 or are we ok going to 1.3.0?

Thats a good question. Technically this is a breaking change (renaming an existing field) but since the SEP is in a draft state I don't think we need to strictly adhere to semantic versioning. @leighmcculloch curious on your thoughts here.

Reviewer: I'd try and keep the SEP version under 1 until it is in an Active state. We have no rules or process for this, but I've found that useful since v0 is a break anything you want version in regards to semantic versioning.

These prices are indicative. The actual price will be calculated at conversion time once the Anchor receives the funds from a User.

The prices returned include any margin that the provider may keep as a service fee, and this margin may vary depending on the directional flow of funds, amount, delivery method, country code, or other factors.

#### Request

Name | Type | Description
-----|------|------------
`sell_asset` | string | The asset you want to sell, using the [Asset Identification Format](#asset-identification-format).
`sell_amount` | string | The amount of `sell_asset` the client would exchange for each of the `buy_assets`.

> `sell_delivery_method` | string | (optional) One of the `name` values specified by the `sell_delivery_methods` array for the associated asset returned from [`GET /info`](#info). Must be provided if the user is making an off-chain deposit to the anchor.

Reviewer 2: Can we avoid the terms `deposit`? That way we limit the number of terms we use when describing the movement of funds.

```suggestion
`sell_delivery_method` | string | (optional) One of the `name` values specified by the `sell_delivery_methods` array for the associated asset returned from [`GET /info`](#info). Must be provided if the user is delivering an off-chain asset to the anchor.
```

Author: Fixed in https://github.com/stellar/stellar-protocol/pull/989/commits/4256e69a463f879755c70b35e3032e28a3643391

> `buy_delivery_method` | string | (optional) One of the `name` values specified by the `buy_delivery_methods` array for the associated asset returned from [`GET /info`](#info). Must be provided if the user intends to receive an off-chain deposit from the anchor.

Reviewer 2: Same thing here, and for the attributes on the `/price` endpoint.

```suggestion
`buy_delivery_method` | string | (optional) One of the `name` values specified by the `buy_delivery_methods` array for the associated asset returned from [`GET /info`](#info). Must be provided if the user intends to receive an off-chain asset from the anchor.
```

Author: Fixed in https://github.com/stellar/stellar-protocol/pull/989/commits/4256e69a463f879755c70b35e3032e28a3643391
